File: server/internal/engine/engine.go
package engine

import (
	"context"
	"errors"
	"log/slog"
	"sync"
	"time"

	"github.com/josephsintum/zed-package-checker/server/internal/model"
)

// DefaultDebounce is how long the engine waits for changes to settle before
// scanning.
//
// A single `npm install` writes the manifest and the lockfile within
// milliseconds of each other, and a `git checkout` can touch dozens of files.
// One second matches what JetBrains uses for the same job.
const DefaultDebounce = time.Second

// Engine schedules scans and tracks what has been published.
//
// Every field below the channels is owned by run() and touched from nowhere
// else. Callers interact only through the methods, which communicate over
// channels, so there is no shared mutable state and therefore no locking.
type Engine struct {
	scanner   Scanner
	publisher Publisher
	log       *slog.Logger
	debounce  time.Duration
	newTimer  func(time.Duration) timer

	requests chan Reason
	queries  chan query
	clears   chan string
	done     chan struct{}
	stopOnce sync.Once
	wg       sync.WaitGroup

	// Owned by run().
	root      string
	report    model.Report
	published map[string]bool
	scanGen   uint64
}

// Option configures an Engine.
type Option func(*Engine)

// WithDebounce sets how long changes must settle before a scan runs.
func WithDebounce(d time.Duration) Option { return func(e *Engine) { e.debounce = d } }

// withTimer replaces the debounce timer, for tests.
func withTimer(f func(time.Duration) timer) Option {
	return func(e *Engine) { e.newTimer = f }
}

// New builds an Engine. Start must be called before it does anything.
func New(log *slog.Logger, scanner Scanner, publisher Publisher, opts ...Option) *Engine {
	e := &Engine{
		scanner:   scanner,
		publisher: publisher,
		log:       log,
		debounce:  DefaultDebounce,
		newTimer:  newRealTimer,
		// Capacity one, with non-blocking sends: a pending request already
		// means "rescan", so extra ones are dropped rather than queued and the
		// channel can never back up.
		requests:  make(chan Reason, 1),
		queries:   make(chan query),
		clears:    make(chan string, 16),
		done:      make(chan struct{}),
		published: make(map[string]bool),
	}
	for _, opt := range opts {
		opt(e)
	}
	return e
}

// Start begins scheduling. It returns immediately; scanning happens in the
// background until ctx is cancelled or Close is called.
func (e *Engine) Start(ctx context.Context, root string) {
	e.root = root
	e.wg.Go(func() { e.run(ctx) })
}

// Close stops the engine and waits for every goroutine it started.
//
// Safe to call more than once, and safe to call without Start.
func (e *Engine) Close() error {
	e.stopOnce.Do(func() { close(e.done) })
	e.wg.Wait()
	return nil
}

// Request asks for a scan. Never blocks.
//
// The request is dropped when one is already pending, which is not a loss:
// a scan that has not started yet will see whatever the filesystem holds when
// it does.
func (e *Engine) Request(reason Reason) {
	select {
	case e.requests <- reason:
	default:
	}
}

// Clear publishes an empty diagnostic set for a path, used when a manifest is
// deleted or renamed. Never blocks.
func (e *Engine) Clear(path string) {
	select {
	case e.clears <- path:
	default:
	}
}

// Findings returns the current findings for a path, or nil.
//
// Used to re-publish when a file is opened, which is how diagnostics survive
// the editor discarding them for a closed buffer. Returns nil if the engine has
// stopped.
func (e *Engine) Findings(path string) []model.Finding {
	reply := make(chan []model.Finding, 1)
	select {
	case e.queries <- query{path: path, reply: reply}:
		return <-reply
	case <-e.done:
		return nil
	}
}

// query is a synchronous read served by run().
type query struct {
	path  string
	reply chan []model.Finding
}

// scanResult carries a completed scan back to the loop.
//
// gen identifies which request produced it, so results from a scan that has
// been superseded can be recognised and discarded. Cancellation is best-effort:
// a scan may finish despite its context being cancelled, and publishing those
// results would show findings for a state of the project that no longer exists.
type scanResult struct {
	gen    uint64
	report model.Report
	err    error
}

// run is the only goroutine that touches engine state.
func (e *Engine) run(ctx context.Context) {
	t := e.newTimer(e.debounce)
	defer t.Stop()

	results := make(chan scanResult, 1)
	var cancelInFlight context.CancelFunc
	cancelPending := func() {
		if cancelInFlight != nil {
			cancelInFlight()
			cancelInFlight = nil
		}
	}
	defer cancelPending()

	for {
		select {
		case reason := <-e.requests:
			// Supersede any scan already running: its results describe a state
			// of the project that has just changed.
			cancelPending()
			e.log.Debug("scan requested", "reason", string(reason))
			t.Reset(e.debounce)

		case <-t.C():
			e.scanGen++
			scanCtx, cancel := context.WithCancel(ctx)
			cancelInFlight = cancel
			gen := e.scanGen

			e.wg.Go(func() {
				report, err := e.scanner.Scan(scanCtx, e.root)
				select {
				case results <- scanResult{gen: gen, report: report, err: err}:
				case <-e.done:
				case <-ctx.Done():
				}
			})

		case res := <-results:
			if res.gen != e.scanGen {
				// Superseded by a newer request while it was running.
				continue
			}
			cancelInFlight = nil
			e.handleResult(ctx, res)

		case path := <-e.clears:
			e.publishOne(ctx, path, nil)
			delete(e.published, path)

		case q := <-e.queries:
			q.reply <- e.findingsFor(q.path)

		case <-ctx.Done():
			return
		case <-e.done:
			return
		}
	}
}

// handleResult records a completed scan and publishes the difference.
func (e *Engine) handleResult(ctx context.Context, res scanResult) {
	if res.err != nil {
		if errors.Is(res.err, context.Canceled) {
			return
		}
		// A failed scan leaves the previous diagnostics in place. Clearing them
		// would suggest the project became clean, which is not what happened.
		e.log.Warn("scan failed", "error", res.err)
		return
	}

	e.report = res.report
	byFile := res.report.ByFile()

	// Named before anything is published, because e.published means "last
	// scan's files" here and "this scan's files" a few lines below.
	previous := e.published

	for path, findings := range byFile {
		e.publishOne(ctx, path, findings)
	}

	// Files that had findings last time and have none now must be published
	// empty, or the editor keeps showing what was fixed.
	for path := range previous {
		if _, still := byFile[path]; !still {
			e.publishOne(ctx, path, nil)
		}
	}

	e.published = make(map[string]bool, len(byFile))
	for path := range byFile {
		e.published[path] = true
	}

	e.log.Info("scan complete",
		"findings", len(res.report.Findings), "files", len(byFile))
}

// publishOne sends one file's diagnostics, logging rather than failing: a
// publish that does not land is not a reason to stop scanning.
func (e *Engine) publishOne(ctx context.Context, path string, findings []model.Finding) {
	if err := e.publisher.Publish(ctx, path, findings); err != nil {
		e.log.Warn("publish failed", "path", path, "error", err)
	}
}

// findingsFor returns the findings anchored at path in the current report.
func (e *Engine) findingsFor(path string) []model.Finding {
	var out []model.Finding
	for _, f := range e.report.Findings {
		if f.AnchorSite().Path == path {
			out = append(out, f)
		}
	}
	return out
}

File: server/src/engine.rs
//! Deciding *when* to scan, and what to publish.
//!
//! One task owns every piece of mutable state; everything else talks to it over
//! a channel — the actor shape, because it is race-free by construction rather
//! than by discipline. Supersession aborts the in-flight scan rather than asking
//! it to notice, so nothing in the scan path polls for cancellation.

use crate::model::{Finding, Report};
use crate::scan::{ScanError, Scanner};
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::{mpsc, oneshot};
use tokio::task::JoinHandle;

/// Matches the `JetBrains` plugin, and is long enough that a `git checkout`
/// touching forty files causes one scan.
pub const DEFAULT_DEBOUNCE: Duration = Duration::from_secs(1);

/// Why a scan was asked for. Reported in logs; it never changes what is done.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Reason {
    /// The server just initialised.
    Startup,
    /// A watched manifest or lockfile changed.
    FileChanged,
    /// The editor saved a manifest.
    FileSaved,
    /// The advisory database landed or changed.
    DatabaseSync,
    /// Configuration changed, or a rescan was asked for.
    Manual,
}

impl Reason {
    /// The reason as it appears in logs.
    pub const fn as_str(self) -> &'static str {
        match self {
            Reason::Startup => "startup",
            Reason::FileChanged => "file changed",
            Reason::FileSaved => "file saved",
            Reason::DatabaseSync => "database updated",
            Reason::Manual => "requested",
        }
    }
}

/// Sends diagnostics to the client. Implemented by the LSP layer.
pub trait Publisher: Send + Sync + 'static {
    /// Replaces a file's diagnostics; an empty set clears them.
    fn publish(&self, path: PathBuf, findings: Vec<Finding>);

    /// Something the user should know that is not a diagnostic.
    ///
    /// Default-empty so a test publisher need not care. Called rarely and at
    /// most once per condition: an editor notification per debounce would be
    /// worse than silence.
    fn notice(&self, _message: String) {}
}

enum Message {
    Request(Reason),
    Clear(PathBuf),
    Query(PathBuf, oneshot::Sender<Vec<Finding>>),
}

/// Asks for a scan without holding the engine.
///
/// The scanner needs to request a rescan when a background download finishes,
/// and the engine needs the scanner — a cycle. The channel exists before either
/// end does, so there is no window in which one side holds an empty handle.
#[derive(Clone)]
pub struct Requester {
    tx: mpsc::Sender<Message>,
}

impl Requester {
    /// Asks for a scan. Never blocks: a full queue already holds the same
    /// instruction.
    pub fn request(&self, reason: Reason) {
        let _ = self.tx.try_send(Message::Request(reason));
    }
}

/// The receiving half, before a scanner and publisher exist to serve it.
pub struct PendingEngine {
    tx: mpsc::Sender<Message>,
    rx: mpsc::Receiver<Message>,
}

impl PendingEngine {
    pub fn start(
        self,
        root: PathBuf,
        scanner: Arc<dyn Scanner>,
        publisher: Arc<dyn Publisher>,
        debounce: Duration,
    ) -> Engine {
        let task = tokio::spawn(run(
            root,
            Arc::clone(&scanner),
            publisher,
            debounce,
            self.rx,
        ));
        Engine {
            tx: self.tx,
            task,
            scanner,
        }
    }
}

/// A handle to the scheduling task.
pub struct Engine {
    tx: mpsc::Sender<Message>,
    task: JoinHandle<()>,
    scanner: Arc<dyn Scanner>,
}

impl Engine {
    /// Creates the request channel before the engine, so something that needs to
    /// ask for a scan can be built first.
    pub fn pending() -> (Requester, PendingEngine) {
        // Senders never block: a pending request already means "rescan", so
        // extra ones are dropped rather than queued and the channel can never
        // back up.
        let (tx, rx) = mpsc::channel(64);
        (Requester { tx: tx.clone() }, PendingEngine { tx, rx })
    }

    /// Starts the scheduler when nothing else needs to ask for scans.
    pub fn start(
        root: PathBuf,
        scanner: Arc<dyn Scanner>,
        publisher: Arc<dyn Publisher>,
        debounce: Duration,
    ) -> Engine {
        let (_, pending) = Engine::pending();
        pending.start(root, scanner, publisher, debounce)
    }

    /// Asks for a scan. Never blocks, and never fails: a full queue already
    /// holds the same instruction.
    pub fn request(&self, reason: Reason) {
        let _ = self.tx.try_send(Message::Request(reason));
    }

    /// Drops a file's diagnostics immediately, for a manifest that was deleted.
    pub fn clear(&self, path: PathBuf) {
        let _ = self.tx.try_send(Message::Clear(path));
    }

    /// The findings already known for a file, for re-publishing on `didOpen`.
    pub async fn findings(&self, path: PathBuf) -> Vec<Finding> {
        let (reply, answer) = oneshot::channel();
        if self.tx.send(Message::Query(path, reply)).await.is_err() {
            return Vec::new();
        }
        answer.await.unwrap_or_default()
    }

    /// Stops the scanner's background work without stopping the scheduler,
    /// for a `shutdown` request that cannot take the engine by value.
    pub fn stop_background(&self) {
        self.scanner.shutdown();
    }

    /// Stops the scheduler and waits for it. Dropping the handle is enough to
    /// stop it; this exists so shutdown is deterministic.
    pub async fn shutdown(self) {
        drop(self.tx);
        let _ = self.task.await;
        self.scanner.shutdown();
    }
}

async fn run(
    root: PathBuf,
    scanner: Arc<dyn Scanner>,
    publisher: Arc<dyn Publisher>,
    debounce: Duration,
    mut rx: mpsc::Receiver<Message>,
) {
    let mut report: HashMap<PathBuf, Vec<Finding>> = HashMap::new();
    let mut published: HashSet<PathBuf> = HashSet::new();
    let mut deadline: Option<tokio::time::Instant> = None;
    let mut in_flight: Option<JoinHandle<Result<Report, ScanError>>> = None;
    // One notice per condition rather than one per debounce.
    let mut announced = false;
    let mut pending = false;
    let mut announced_partial = false;

    loop {
        // `sleep_until` on a deadline rather than a resettable timer: setting a
        // new deadline *is* the reset, so there is no timer to drain.
        let debounce_elapsed = async {
            match deadline {
                Some(at) => tokio::time::sleep_until(at).await,
                None => std::future::pending().await,
            }
        };
        let scan_finished = async {
            match in_flight.as_mut() {
                Some(handle) => handle.await,
                None => std::future::pending().await,
            }
        };

        tokio::select! {
            message = rx.recv() => {
                let Some(message) = message else { break };
                match message {
                    Message::Request(reason) => {
                        tracing::debug!(reason = reason.as_str(), "scan requested");
                        // A newer request supersedes whatever is running: its
                        // results describe a tree that no longer exists.
                        if let Some(handle) = in_flight.take() {
                            handle.abort();
                        }
                        deadline = Some(tokio::time::Instant::now() + debounce);
                    }
                    Message::Clear(path) => {
                        report.remove(&path);
                        if published.remove(&path) {
                            publisher.publish(path, Vec::new());
                        }
                    }
                    Message::Query(path, reply) => {
                        let _ = reply.send(report.get(&path).cloned().unwrap_or_default());
                    }
                }
            }
            () = debounce_elapsed => {
                deadline = None;
                let scanner = Arc::clone(&scanner);
                let root = root.clone();
                // Scanning parses archives and walks a tree; it belongs on the
                // blocking pool, not on a runtime worker.
                in_flight = Some(tokio::task::spawn_blocking(move || scanner.scan(&root)));
            }
            finished = scan_finished => {
                in_flight = None;
                match finished {
                    Ok(Ok(fresh)) => {
                        // Said once, the first time something leaves the
                        // machine, rather than on every scan.
                        if let crate::model::Source::Api { checked } = fresh.source
                            && !announced
                        {
                            announced = true;
                            publisher.notice(format!(
                                "Checked {checked} {} against osv.dev — names and versions only. \
                                 Set online.enabled to false to use the offline database instead.",
                                if checked == 1 { "dependency" } else { "dependencies" }
                            ));
                        }
                        // An unchecked ecosystem produces a report that looks
                        // exactly like a clean one. Say which, once.
                        if let crate::model::Source::PartialArchive { missing } = &fresh.source
                            && !announced_partial
                        {
                            announced_partial = true;
                            publisher.notice(format!(
                                "Not checked yet: {}. Still downloading those advisories.",
                                missing
                                    .iter()
                                    .map(ToString::to_string)
                                    .collect::<Vec<_>>()
                                    .join(", ")
                            ));
                        }
                        pending = false;
                        publish(&fresh, &mut report, &mut published, publisher.as_ref());
                    }
                    Ok(Err(error)) => {
                        // "Still downloading" and "nothing is wrong" must not
                        // look alike, and an empty diagnostic set says the
                        // second. Announced once per streak, not per debounce.
                        if !pending {
                            pending = true;
                            publisher.notice(format!("Dependencies not checked yet: {error}"));
                        }
                        tracing::warn!(%error, "scan failed");
                    }
                    // Cancelled by a newer request, which is the normal path.
                    Err(join) if join.is_cancelled() => {}
                    Err(join) => tracing::warn!(error = %join, "scan task failed"),
                }
            }
        }
    }

    if let Some(handle) = in_flight {
        handle.abort();
    }
}

/// Publishes a fresh report, including empty sets for files that were affected
/// last time and are not any more.
fn publish(
    fresh: &Report,
    report: &mut HashMap<PathBuf, Vec<Finding>>,
    published: &mut HashSet<PathBuf>,
    publisher: &dyn Publisher,
) {
    let grouped = fresh.by_file();
    let current: HashMap<PathBuf, Vec<Finding>> = grouped
        .into_iter()
        .map(|(path, findings)| (path.to_path_buf(), findings.into_iter().cloned().collect()))
        .collect();

    // Clearing comes first so a file that moved from "vulnerable" to "clean"
    // never keeps a stale squiggle, even briefly.
    for stale in published.difference(&current.keys().cloned().collect()) {
        publisher.publish(stale.clone(), Vec::new());
    }

    *published = current.keys().cloned().collect();
    for (path, findings) in &current {
        publisher.publish(path.clone(), findings.clone());
    }
    *report = current;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::{Ecosystem, Fix, Package, Range, Site};
    use std::path::Path;
    use std::sync::Mutex;
    use std::sync::atomic::{AtomicUsize, Ordering};

    /// Short enough to keep the tests quick, long enough that several requests
    /// land inside one window on a loaded machine.
    const DEBOUNCE: Duration = Duration::from_millis(60);

    fn finding(path: &str) -> Finding {
        Finding {
            package: Package::new(Ecosystem::Npm, "lodash", "4.17.15"),
            advisories: vec![Arc::new(crate::model::Advisory {
                id: "GHSA-1".into(),
                aliases: Box::default(),
                summary: Box::default(),
                cvss_score: 7.5,
                cvss_vector: Box::default(),
                affected: Box::default(),
                references: Box::default(),
            })],
            evidence: Site::new(path, Range::whole_line(1)),
            declared: None,
            paths: Vec::new(),
            from_range: false,
            dep_groups: Vec::new(),
            fix: Fix::None,
        }
    }

    #[derive(Default)]
    struct FakeScanner {
        calls: AtomicUsize,
        findings: Mutex<Vec<Finding>>,
        /// Held while a scan is "running", so a test can supersede one.
        block: Mutex<Option<std::sync::mpsc::Receiver<()>>>,
        /// When set, every scan fails.
        failing: std::sync::atomic::AtomicBool,
    }

    impl Scanner for Arc<FakeScanner> {
        fn scan(&self, root: &Path) -> Result<Report, ScanError> {
            self.calls.fetch_add(1, Ordering::SeqCst);
            if let Some(gate) = self.block.lock().unwrap().take() {
                let _ = gate.recv();
            }
            if self.failing.load(Ordering::SeqCst) {
                return Err(ScanError::NotReady);
            }
            Ok(Report::new(root, self.findings.lock().unwrap().clone()))
        }
    }

    #[derive(Default)]
    struct Recorder {
        published: Mutex<Vec<(PathBuf, usize)>>,
    }

    impl Publisher for Arc<Recorder> {
        fn publish(&self, path: PathBuf, findings: Vec<Finding>) {
            self.published.lock().unwrap().push((path, findings.len()));
        }
    }

    fn harness(findings: Vec<Finding>) -> (Arc<FakeScanner>, Arc<Recorder>, Engine) {
        let scanner = Arc::new(FakeScanner::default());
        *scanner.findings.lock().unwrap() = findings;
        let recorder = Arc::new(Recorder::default());
        let engine = Engine::start(
            PathBuf::from("/project"),
            Arc::new(Arc::clone(&scanner)),
            Arc::new(Arc::clone(&recorder)),
            DEBOUNCE,
        );
        (scanner, recorder, engine)
    }

    /// Long enough for the debounce to elapse and a scan to publish. The
    /// tests run with time paused, so this costs nothing real and is not a
    /// race against the machine.
    async fn settle() {
        tokio::time::sleep(DEBOUNCE * 5).await;
    }

    #[tokio::test(start_paused = true)]
    async fn many_requests_inside_the_window_cause_one_scan() {
        let (scanner, _, engine) = harness(vec![finding("/project/package.json")]);
        for _ in 0..40 {
            engine.request(Reason::FileChanged);
        }
        settle().await;
        assert_eq!(scanner.calls.load(Ordering::SeqCst), 1);
        engine.shutdown().await;
    }

    #[tokio::test(start_paused = true)]
    async fn a_report_is_published_per_file() {
        let (_, recorder, engine) = harness(vec![
            finding("/project/package.json"),
            finding("/project/go.mod"),
        ]);
        engine.request(Reason::Startup);
        settle().await;

        let published = recorder.published.lock().unwrap().clone();
        assert_eq!(published.len(), 2);
        assert!(published.iter().all(|(_, n)| *n == 1));
        engine.shutdown().await;
    }

    #[tokio::test(start_paused = true)]
    async fn a_file_that_becomes_clean_is_published_empty() {
        let (scanner, recorder, engine) = harness(vec![finding("/project/package.json")]);
        engine.request(Reason::Startup);
        settle().await;

        // Nothing vulnerable this time: the previous squiggle has to go.
        scanner.findings.lock().unwrap().clear();
        engine.request(Reason::FileChanged);
        settle().await;

        let published = recorder.published.lock().unwrap().clone();
        assert_eq!(
            published.last(),
            Some(&(PathBuf::from("/project/package.json"), 0)),
            "the last publish must clear the file"
        );
        engine.shutdown().await;
    }

    #[tokio::test(start_paused = true)]
    async fn a_deleted_manifest_is_cleared_without_waiting_for_a_scan() {
        let (_, recorder, engine) = harness(vec![finding("/project/package.json")]);
        engine.request(Reason::Startup);
        settle().await;
        recorder.published.lock().unwrap().clear();

        engine.clear(PathBuf::from("/project/package.json"));
        tokio::time::sleep(DEBOUNCE).await;
        assert_eq!(
            recorder.published.lock().unwrap().clone(),
            [(PathBuf::from("/project/package.json"), 0)]
        );
        engine.shutdown().await;
    }

    #[tokio::test(start_paused = true)]
    async fn findings_are_readable_without_rescanning() {
        let (scanner, _, engine) = harness(vec![finding("/project/package.json")]);
        engine.request(Reason::Startup);
        settle().await;

        let found = engine
            .findings(PathBuf::from("/project/package.json"))
            .await;
        assert_eq!(found.len(), 1);
        assert!(
            engine
                .findings(PathBuf::from("/project/go.mod"))
                .await
                .is_empty()
        );
        // A query must never trigger work.
        assert_eq!(scanner.calls.load(Ordering::SeqCst), 1);
        engine.shutdown().await;
    }

    // Real time: the paused clock does not advance while a blocking scan is
    // held open, and this test holds one open on purpose.
    #[tokio::test]
    async fn a_superseded_scan_never_publishes() {
        let scanner = Arc::new(FakeScanner::default());
        *scanner.findings.lock().unwrap() = vec![finding("/project/package.json")];
        let (release, gate) = std::sync::mpsc::channel::<()>();
        *scanner.block.lock().unwrap() = Some(gate);

        let recorder = Arc::new(Recorder::default());
        let engine = Engine::start(
            PathBuf::from("/project"),
            Arc::new(Arc::clone(&scanner)),
            Arc::new(Arc::clone(&recorder)),
            DEBOUNCE,
        );

        engine.request(Reason::Startup);
        tokio::time::sleep(DEBOUNCE * 2).await; // the first scan is now stuck
        engine.request(Reason::FileChanged); // supersede it
        settle().await;
        drop(release); // let the abandoned scan finish

        settle().await;
        let published = recorder.published.lock().unwrap().clone();
        assert_eq!(
            published.len(),
            1,
            "only the winning scan publishes, got {published:?}"
        );
        engine.shutdown().await;
    }

    #[tokio::test(start_paused = true)]
    async fn a_failed_scan_keeps_the_previous_diagnostics() {
        // Clearing on failure would tell the user the project became clean,
        // which is not what a failed scan means.
        let (scanner, recorder, engine) = harness(vec![finding("/project/package.json")]);
        engine.request(Reason::Startup);
        settle().await;
        assert_eq!(recorder.published.lock().unwrap().len(), 1);

        scanner.failing.store(true, Ordering::SeqCst);
        engine.request(Reason::FileChanged);
        settle().await;

        assert_eq!(
            scanner.calls.load(Ordering::SeqCst),
            2,
            "the failing scan ran"
        );
        assert_eq!(
            recorder.published.lock().unwrap().len(),
            1,
            "a failed scan published something"
        );
        assert_eq!(
            engine
                .findings(PathBuf::from("/project/package.json"))
                .await
                .len(),
            1,
            "findings were dropped on failure"
        );
        engine.shutdown().await;
    }

    // Real time: the paused clock does not advance while a blocking scan is
    // held open, and this test holds one open on purpose.
    #[tokio::test]
    async fn shutdown_returns_while_a_scan_is_in_flight() {
        let scanner = Arc::new(FakeScanner::default());
        let (release, gate) = std::sync::mpsc::channel::<()>();
        *scanner.block.lock().unwrap() = Some(gate);
        let engine = Engine::start(
            PathBuf::from("/project"),
            Arc::new(Arc::clone(&scanner)),
            Arc::new(Arc::new(Recorder::default())),
            DEBOUNCE,
        );

        engine.request(Reason::Startup);
        tokio::time::sleep(DEBOUNCE * 2).await; // the scan is now stuck
        assert_eq!(scanner.calls.load(Ordering::SeqCst), 1);

        let stopped = tokio::time::timeout(Duration::from_secs(2), engine.shutdown()).await;
        drop(release);
        assert!(stopped.is_ok(), "shutdown hung with a scan in flight");
    }
}
